src/bin.rs

```diff
@@ -1,508 +1,877 @@
+            doc
+        })
+    )]
+    Palette(#[bpaf(external(palette_args))] PaletteArgs),
+    /// Generate shell completions (zsh, bash, fish, and elvish)
+    #[bpaf(command)]
+    Completions {
+        /// Shell to generate completions for.
+        #[bpaf(
+            positional("SHELL"),
+            guard(
+                |s| ["zsh", "bash", "fish", "elvish"].contains(&s.to_lowercase().as_str()),
+                "unknown shell"
+            )
+        )]
+        shell: String,
+    },
+}
 
-        name
-    }
+fn load_image<P: AsRef<Path>>(path: P) -> Result<Image, String> {
+    let path = path.as_ref();
+    let time = Instant::now();
+    let lut = image::io::Reader::open(path)
+        .map_err(|e| format!("failed to open image: {e}"))?
+        .with_guessed_format()
+        .map_err(|e| format!("failed to guess image format: {e}"))?
+        .decode()
+        .map_err(|e| format!("failed to decode image: {e}"))?
+        .to_rgb8();
+    println!("✔ Loaded {path:?} in {:.2?}", time.elapsed());
+    Ok(lut)
+}
 
-    fn detail_string(&self) -> String {
-        let mut buf = format!(
-            "hald{}_{:?}_lum{}",
-            self.level, self.algorithm, self.luminosity
-        );
-        match self.algorithm {
-            Algorithm::GaussianSampling => buf.push_str(&format!(
-                "_{}_{}_{}",
-                self.mean, self.std_dev, self.iterations
-            )),
-            Algorithm::ShepardsMethod => {
-                buf.push_str(&format!("_pow{}_near{}", self.power, self.nearest));
-            },
-            Algorithm::GaussianRBF => {
-                buf.push_str(&format!("_shape{}_near{}", self.shape, self.nearest));
+impl Lutgen {
+    fn execute(self) -> Result<String, String> {
+        match self {
+            Lutgen::Generate {
+                output,
+                palette,
+                lut_algorithm,
+                extra_colors,
+            } => Lutgen::generate(output, palette, lut_algorithm, extra_colors),
+            Lutgen::Apply {
+                dir,
+                ref output,
+                ref palette,
+                cache,
+                ref hald_clut_or_algorithm,
+                ref input,
+                ref extra_colors,
+            } => {
+                let hash =
+                    if cache && !matches!(hald_clut_or_algorithm, LutAlgorithm::HaldClut { .. }) {
+                        let mut hasher = DefaultHasher::new();
+                        self.hash(&mut hasher);
+                        Some(hasher.finish())
+                    } else {
+                        None
+                    };
+
+                Lutgen::apply(
+                    hash,
+                    dir,
+                    output.to_owned(),
+                    palette.to_owned(),
+                    hald_clut_or_algorithm.to_owned(),
+                    input.to_owned(),
+                    extra_colors.to_owned(),
+                )
```

Comment:
you could the `this @ Lutgen::Apply { .. }` syntax here then use `this.hash(&mut hasher)` and after restructure `this`. So that `.to_owned` calls could be removed.